File: 322. Coin Change/322. Coin Change top-down.java
class Solution {
    private int[] coins, count;

    public int coinChange(int[] coins, int amount) {
        this.coins = coins;
        this.count = new int[amount + 1];
        return dp(amount);
    }

    private int dp(int remainder) {
        if (remainder == 0)
            return 0;
        if (remainder < 0)
            return -1;
        if (count[remainder] != 0)
            return count[remainder];

        int min = Integer.MAX_VALUE;
        for (int coin : coins) {
            int res = dp(remainder - coin);
            if (res >= 0 && res < min) {
                min = 1 + res;
            }
        }
        if (min == Integer.MAX_VALUE) {
            count[remainder] = -1;
        } else {
            count[remainder] = min;
        }
        return count[remainder];
    }
}

File: 322. Coin Change/322. Coin Change top-down.py
from cgitb import reset
from enum import Flag
from functools import lru_cache
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        @lru_cache(None)
        def dp(i: int) -> int:
            # base case
            if i == 0:
                return 0
            if i < 0:
                return -1

            # recurrence
            minimum = float('inf')
            for coin in coins:
                res = dp(i-coin)
                if res >= 0 and res < minimum:
                    minimum = 1 + res
            if minimum == float('inf'):
                return -1
            return minimum
        return dp(amount)


s = Solution()
# coins = [1, 2, 5]
# amount = 11
# coins = [3]
# amount = 4
coins = [1, 2147483647]
amount = 2
print(s.coinChange(coins, amount))
